include/boost/json/impl/monotonic_resource.ipp

```diff
@@ -10,61 +11,113 @@
 #ifndef BOOST_JSON_IMPL_MONOTONIC_RESOURCE_IPP
 #define BOOST_JSON_IMPL_MONOTONIC_RESOURCE_IPP
 
+#include <algorithm>
+
 #include <boost/json/monotonic_resource.hpp>
 
 namespace boost {
 namespace json {
 
-struct monotonic_resource::block
+// ensures that the alignment of top is
+// the strictest fundamental alignment requirement
+struct alignas(alignof(std::max_align_t)) 
+    monotonic_resource::block
 {
     std::size_t const size;
+    block* next = nullptr;
     std::uintptr_t top;
-    block* next;
 
     block(
         std::size_t size_,
         block* next_)
-        : size(size_)
-        , top(reinterpret_cast<
-            std::uintptr_t>(this+1))
-        , next(next_)
+        : size(size_), 
+          next(next_),
+          top(base())
     {
     }
 
-    void*
-    alloc(
-        std::size_t n,
-        std::size_t align) noexcept
+    std::uintptr_t
+    base() const
     {
-        // must be power of 2
-        BOOST_ASSERT(
-            (align & (align - 1)) == 0);
-        auto i = top;
-        if((i & (align - 1)) != 0)
-            i = (i | (align - 1)) + 1;
-        if(i + n > (reinterpret_cast<
-            std::uintptr_t>(this+1) + size))
-            return nullptr;
-        top = i + n;
-        return reinterpret_cast<void*>(i);
+        return reinterpret_cast<
+            std::uintptr_t>(this + 1);
     }
 };
 
+std::uintptr_t
+monotonic_resource::
+align_address(
+    std::uintptr_t addr,
+    std::size_t align)
+{
+    // must be power of 2
+    BOOST_ASSERT(
+        (align & (align - 1)) == 0);
+    return addr & (align - 1) ?
+        (addr | (align - 1)) + 1 : addr;
+}
+
+template<typename Block>
+void*
+monotonic_resource::
+allocate_in_block(
+    Block& blk,
+    std::size_t size,
+    std::size_t align)
+{
+    auto aligned_addr =
+        align_address(blk.top, align);
+    if(aligned_addr + size > blk.base() + blk.size)
```

Comment:
you should structure the LIKELY case to be true. This condition will be false more often than it is true, thus you should invert it:
```
if(aligned_addr + size <= blk.base() + blk.size)
{
    blk.top = aligned_addr + size;
    return reinterpret_cast<void*>(aligned_addr);
}
return nullptr;
```